File: 월간 코드 챌린지/괄호 회전하기.py
from collections import deque
def correct(x):
    stack = []
    result = True
    for i in x:
        if i == '[' or i == '(' or i == '{':
            stack.append(i)
        else:
            if stack == []:
                result = False
                break
            else:
                if i == ']':
                    if stack[-1] == '[':
                        stack.pop(-1)
                    else:
                        result = False
                        break
                elif i == ')':
                    if stack[-1] == '(':
                        stack.pop(-1)
                    else:
                        result = False
                        break                        
                elif i == '}':
                    if stack[-1] == '{':
                        stack.pop(-1)   
                    else:
                        result = False
                        break
    if stack:
        result = False
    return result                        
                
def solution(s):
    answer = 0
    q = deque(s)
    cnt = 0
    tmp = []
    while cnt < len(s):
        q.rotate(-1)
        if correct(q):
            answer += 1
        cnt += 1   
    return answer








-------------------------------------------
from collections import deque
def correct(x):
    stack = []
    result = True
    for i in x:
        if i == '[' or i == '(' or i == '{':
            stack.append(i)
        else:
            if stack == []:
                result = False
                break
            else:
                if i == ']':
                    if stack[-1] == '[':
                        stack.pop(-1)
                    else:
                        result = False
                        break
                elif i == ')':
                    if stack[-1] == '(':
                        stack.pop(-1)
                    else:
                        result = False
                        break                        
                elif i == '}':
                    if stack[-1] == '{':
                        stack.pop(-1)   
                    else:
                        result = False
                        break
    if stack:
        result = False
    return result    

def solution(s):
    answer = 0
    for i in range(len(s)):
        answer += correct(s[i:]+s[:i])
    return answer    

File: Level 2/괄호 회전하기.java
// https://school.programmers.co.kr/learn/courses/30/lessons/76502
import java.util.*;
class Solution {
    public int solution(String s) {
        int answer = 0;
       
        for (int i=0; i<s.length(); i++) {
            if (isCorrect(s.substring(i) + s.substring(0, i))) {
                answer++;
            }
        }
        return answer;
    }
    
    
    public boolean isCorrect(String s) {
        
        boolean result = true;
        Stack<Character> stack = new Stack<>();
        
        for (int i=0; i<s.length(); i++) {
            if ((s.charAt(i) == '(') || (s.charAt(i) == '[') || (s.charAt(i) == '{')) {
                stack.push(s.charAt(i));
            } else {
                if (stack.size() == 0) {
                    result = false;
                    break;
                } else {
                    if ((s.charAt(i) == ')') && (stack.peek() == '(')) {
                        stack.pop();
                    } else if ((s.charAt(i) == ']') && (stack.peek() == '[')) {
                        stack.pop();
                    } else if ((s.charAt(i) == '}') && (stack.peek() == '{')) {
                        stack.pop();
                    }
                }   
            }
        }
        
        if (!stack.isEmpty()) {
            result = false;
        }
        return result;
    }
    
}
